openstack/identity/v3/extensions/ec2tokens/requests.go

```diff
@@ -0,0 +1,279 @@
+	// token expires.
+	AllowReauth bool `json:"-"`
+	// Can be either a []byte (encoded to base64 automatically) or a
+	// string. You can set the singature explicitly, when you already know
+	// it.
+	Signature interface{} `json:"signature"`
+	// Optional random hash, when signature is nil.
+	BodyHash *string `json:"body_hash"`
+	// Optional timestamp to calculate a V4 signature.
+	Timestamp *time.Time `json:"-"`
+}
+
+func canonicalQsV2(params map[string]string) string {
+	var keys []string
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	var pairs []string
+	for _, k := range keys {
+		pairs = append(pairs, fmt.Sprintf("%s=%s", k, url.QueryEscape(params[k])))
+	}
+
+	return strings.Join(pairs, "&")
+}
+
+func stringToSignV2(opts AuthOptions, params map[string]string) string {
+	stringToSign := strings.Join([]string{
+		opts.Verb,
+		opts.Host,
+		opts.Path,
+	}, "\n")
+
+	return strings.Join([]string{
+		stringToSign,
+		canonicalQsV2(params),
+	}, "\n")
+}
+
+func canonicalQsV4(verb string, params map[string]string) string {
+	if verb == "POST" {
+		return ""
+	}
+	return canonicalQsV2(params)
+}
+
+func canonicalHeadersV4(headers map[string]string, signedHeaders string) string {
+	headersLower := make(map[string]string, len(headers))
+	for k, v := range headers {
+		headersLower[strings.ToLower(k)] = v
+	}
+
+	var headersList []string
+	for _, h := range strings.Split(signedHeaders, ";") {
+		if v, ok := headersLower[h]; ok {
+			headersList = append(headersList, h+":"+v)
+		}
+	}
+
+	return strings.Join(headersList, "\n") + "\n"
+}
+
+func sumHMAC1(key []byte, data []byte) []byte {
+	hash := hmac.New(sha1.New, key)
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
+func sumHMAC256(key []byte, data []byte) []byte {
+	hash := hmac.New(sha256.New, key)
+	hash.Write(data)
+	return hash.Sum(nil)
+}
+
+func signatureKeyV4(secret, date, region, service string) []byte {
+	kDate := sumHMAC256([]byte("AWS4"+secret), []byte(date))
+	kRegion := sumHMAC256(kDate, []byte(region))
+	kService := sumHMAC256(kRegion, []byte(service))
+	return sumHMAC256(kService, []byte(aws4Request))
+}
+
+func signatureV4(opts AuthOptions, params map[string]string, date time.Time, bodyHash string) string {
+	scope := strings.Join([]string{
+		date.Format(yyyymmdd),
+		opts.Region,
+		opts.Service,
+		aws4Request,
+	}, "/")
+
+	canonicalRequest := strings.Join([]string{
+		opts.Verb,
+		opts.Path,
+		canonicalQsV4(opts.Verb, params),
+		canonicalHeadersV4(opts.Headers, params["X-Amz-SignedHeaders"]),
+		params["X-Amz-SignedHeaders"],
+		bodyHash,
+	}, "\n")
+	hash := sha256.Sum256([]byte(canonicalRequest))
+
+	strToSign := strings.Join([]string{
+		v4Algo,
+		date.Format(iso8601utc),
+		scope,
+		hex.EncodeToString(hash[:]),
+	}, "\n")
+
+	key := signatureKeyV4(opts.Secret, date.Format(yyyymmdd), opts.Region, opts.Service)
+
+	return hex.EncodeToString(sumHMAC256(key, []byte(strToSign)))
+}
+
+func randomBodyHash() string {
```

Comment:
This function seems simple enough that I think we can forgo renaming, doc strings, unit tests, etc. However, I recommend moving this, and the other helper functions, to the very bottom of this file to denote that it's more of a simple helper function along with a general comment saying something along the lines of "The following are small helper functions used to help build the signature".

Reply:
done